pyop2/global_kernel.py

```diff
@@ -0,0 +1,387 @@
+        same for each base entity?
+    :param subset: Are we iterating over a subset?
+    :param iteration_region: :class:`IterationRegion` representing the set of
+        entities being iterated over. Only valid if looping over an extruded mesh.
+        Valid values are:
+          - ``ON_BOTTOM``: iterate over the bottom layer of cells.
+          - ``ON_TOP`` iterate over the top layer of cells.
+          - ``ALL`` iterate over all cells (the default if unspecified)
+          - ``ON_INTERIOR_FACETS`` iterate over all the layers
+             except the top layer, accessing data two adjacent (in
+             the extruded direction) cells at a time.
+    :param pass_layer_arg: Should the wrapper pass the current layer into the
+        kernel (as an `int`). Only makes sense for indirect extruded iteration.
+    """
+
+    _cppargs = []
+    _libraries = []
+    _system_headers = []
+
+    _cache = {}
+
+    @classmethod
+    def _cache_key(cls, local_knl, arguments, **kwargs):
+        key = [cls, local_knl.cache_key,
+               *kwargs.items(), configuration["simd_width"]]
+
+        key.extend([a.cache_key for a in arguments])
+
+        counter = itertools.count()
+        seen_maps = collections.defaultdict(lambda: next(counter))
+        key.extend([seen_maps[m] for a in arguments for m in a.maps])
+
+        return tuple(key)
+
+    def __init__(self, local_kernel, arguments, *,
+                 extruded=False,
+                 constant_layers=False,
+                 subset=False,
+                 iteration_region=None,
+                 pass_layer_arg=False):
+        if self._initialized:
+            return
+
+        if not len(local_kernel.accesses) == len(arguments):
+            raise ValueError("Number of arguments passed to the local "
+                             "and global kernels do not match")
+
+        if pass_layer_arg and not extruded:
+            raise ValueError("Cannot request layer argument for non-extruded iteration")
+        if constant_layers and not extruded:
+            raise ValueError("Cannot request constant_layers argument for non-extruded iteration")
+
+        self.local_kernel = local_kernel
+        self.arguments = arguments
+        self._extruded = extruded
+        self._constant_layers = constant_layers
+        self._subset = subset
+        self._iteration_region = iteration_region
+        self._pass_layer_arg = pass_layer_arg
+
+        # Cache for stashing the compiled code
+        self._func_cache = {}
+
+        self._initialized = True
+
+    @mpi.collective
+    def __call__(self, comm, *args):
+        """Execute the compiled kernel.
+
+        :arg comm: Communicator the execution is collective over.
+        :*args: Arguments to pass to the compiled kernel.
+        """
+        # If the communicator changes then we cannot safely use the in-memory
+        # function cache.
+        key = id(mpi.dup_comm(comm))
```

Comment:
This dup_comm is unnecessary, since the comm coming in is attached to an iterset and has already been dupped.

Reply:
Ah fantastic. This line was the reason for the small performance regression we were seeing so it's nice that it's not necessary. Will fix.